.github/workflows/linux.yml

```diff
@@ -30,17 +30,18 @@ jobs:
         submodules: true
     - name: Fixup Run actions/checkout
       run: git config --global --add safe.directory '*'
+    - name: Setup oneapi environment
+      run: |
+        source /opt/intel/oneapi/setvars.sh
+        printenv >> $GITHUB_ENV
```

Comment:
The real problem is that `/opt/intel/oneapi/compiler/latest/linux/compiler/lib/intel64_lin/` (the location of `libiomp5.so`) is missing from RPATH.

```
$ readelf -d build/pairinteraction/libpicomplex.so

Dynamic section at offset 0x7d4670 contains 38 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libsqlite3.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgsl.so.25]
 0x0000000000000001 (NEEDED)             Shared library: [libgomp.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_intel_lp64.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_intel_thread.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_core.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libiomp5.so]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000000000000e (SONAME)             Library soname: [libpicomplex.so]
 0x000000000000001d (RUNPATH)            Library runpath: [/opt/intel/oneapi/mkl/2023.1.0/lib/intel64]
 0x000000000000000c (INIT)               0x4aa000
 0x000000000000000d (FINI)               0x6f4d00
 0x0000000000000019 (INIT_ARRAY)         0x7d4428
 0x000000000000001b (INIT_ARRAYSZ)       296 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x7d4550
 0x000000000000001c (FINI_ARRAYSZ)       152 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x298
 0x0000000000000005 (STRTAB)             0x8d660
 0x0000000000000006 (SYMTAB)             0x253f8
 0x000000000000000a (STRSZ)              2188688 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x7d6000
 0x0000000000000002 (PLTRELSZ)           335280 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x458030
 0x0000000000000007 (RELA)               0x2ac8c0
 0x0000000000000008 (RELASZ)             1750896 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0x2ac6d0
 0x000000006fffffff (VERNEEDNUM)         7
 0x000000006ffffff0 (VERSYM)             0x2a3bf0
 0x000000006ffffff9 (RELACOUNT)          72308
 0x0000000000000000 (NULL)               0x0
```

Reply:
> Please don't do this. Making whether it works dependent on the presence of random environment variables is not good and will just cause confusion.

To my understanding, oneapi is meant to be used in an environment where the environment variables are set with `. /opt/intel/oneapi/setvars.sh`. I am not sure whether this is the best solution and whether one can use it to create packages that are portable to other systems. Still, it seems to work in commit [d77438d](https://github.com/pairinteraction/pairinteraction/pull/147/commits/d77438d5123e5b71f5d6a81469b44decdb643a50). For me, it would be fine to use this solution and think about providing packages that make use of mkl at a later point. Then, for the moment, linking against mkl would be an option for advanced users that compile pairinteraction on their own.

Reply:
I'd say, revert all of that and go with `MKL_THREADING="sequential"` for now until we find an *actual* solution.

Reply:
I wonder whether there is a nice solution. Manually modifying the RPATH would also be hacky. If I search for the error or how oneapi is used in other github actions, the standard answer seems to be to run `. /opt/intel/oneapi/setvars.sh`. For example, [pybind11](https://github.com/pybind/pybind11/blob/da91926295638b2b1c0c85a568a2d66ca3646565/.github/workflows/ci.yml) is running it multiple times.

Reply:
> To my understanding, oneapi is meant to be used in an environment where the environment variables are set

I would agree if we were talking about a system that pairinteraction was *not* compiled on, but this is not the case here. The point is that `MKLConfig.cmake` is buggy and needs to be fixed.

Reply:
Hmm, but this is nothing we can fix easily, isn't it? Then, I would prefer to have the `setvars.sh` workaround. I find it handy to have the possibility to use the multithreaded version of mkl, especially as I am currently looking into benchmarking different eigensolver implementations.

Reply:
I fixed it already, see https://github.com/pairinteraction/pairinteraction/pull/147/commits/73251d8657cbb08d443aac4362b876ff86bd7058 and I will submit a bug report upstream once I find out where to do that. `MKLConfig.cmake` sets a variable `OMP_LIBRARY` which contains the full path to `libiomp5`, so all that's left is to extract the directory, dereference all the symlinks, and add it to RPATH.

Reply:
Oh, I just saw it. I am fine with this solution (except for the way `set(MKL_THREADING "intel_thread" CACHE STRING "MKL threading backend")` is called - as written previously, I would prefer to let MKLConfig.cmake determine a threading backend whenever possible). However, somehow, your force push destroyed some bug fixes that I did to other files.

Reply:
I just squashed some of these fixes into https://github.com/pairinteraction/pairinteraction/pull/147/commits/6f19346af3be2a3be3cfd23f89d71d3400ecd259

About https://github.com/pairinteraction/pairinteraction/commit/65c14b08a684f4d0d93159e06e1a8cb9b264da1b I'm not so sure. Seems a bit excessive to set these environment variables for all tests. This might shadow problems arising from multithreaded LAPACK. Maybe only for the `python.parallelization` test would suffice, or just set it from within the test. A call to `os.environ` should then precede `import pairinteraction`.

Reply:
Ok, I completely agree that https://github.com/pairinteraction/pairinteraction/commit/65c14b08a684f4d0d93159e06e1a8cb9b264da1b is a bit excessive. The cleanest solution would be to neither set these environment variables with `set_tests_properties` nor within the `.yml` files but for each individual test. However, I don't know how straightforward it is to figure out which tests require these environment variables and which do not. Without `MKL_NUM_THREADS=1`, surprisingly many tests failed. I just added a `FIXME` to the code as a reminder, see https://github.com/pairinteraction/pairinteraction/pull/147/commits/14d74feb57dcfa753ba16e67fa974ca23e097b39. For me, this would be fine for the moment, but feel free to investigate it in more detail.

Reply:
Strangely, calculated eigenenergies, etc. are different if multiple threads are used. Maybe, this is just because of a different order of the eigenvalues. But we should look into it before merging the pull request to exclude that a user obtains wrong results if the user does not set the environment variables.